mysky-seed.go

```diff
@@ -0,0 +1,279 @@
+// along the way.
+func PhraseToSeed(phrase string) (seed Seed, err error) {
+	// Get the words from the seed.
+	words := strings.Split(phrase, " ")
+	if len(words) != 15 {
+		return seed, errors.New("seed phrase must be 15 words")
+	}
+
+	// Get the index of each word.
+	var wordInts [13]uint16
+	for i := 0; i < 13; i++ {
+		// Figure out which word this is.
+		var j uint16
+		for j = 0; j < uint16(len(myskyDictionary)); j++ {
+			if myskyDictionary[j][:3] == words[i][:3] {
+				break
+			}
+		}
+		// Check that the word exists.
+		if myskyDictionary[j][:3] != words[i][:3] {
+			return seed, errors.New("invalid seed word")
+		}
+		// Check that the word is not overlapping the version bits.
+		if i == 12 && j > 255 {
+			return seed, errors.New("13th word of seed phrase is invalid")
+		}
+		wordInts[i] = j
+	}
+
+	// Convert the words into seed. This is copied from the skynet-js
+	// implementation.
+	curByte := 0
+	curBit := 0
+	for i := 0; i < 13; i++ {
+		word := wordInts[i]
+		wordBits := 10
+		if i == 12 {
+			wordBits = 8
+		}
+		for j := 0; j < wordBits; j++ {
+			bitSet := (word & (1 << (wordBits - j - 1))) > 0
+			if bitSet {
+				seed[curByte] |= 1 << (8 - curBit - 1)
+			}
+			curBit++
+			if curBit >= 8 {
+				curByte++
+				curBit = 0
+			}
+		}
+	}
+
+	// Verify the checksum of the phrase. This is copied from the skynet-js
+	// implementation.
+	checksum := sha512.Sum512(seed[:])
+	word1 := uint32(checksum[0]) << 8 // shift first word to leave room for second
+	word1 += uint32(checksum[1])      // load seconds word
+	word1 >>= 6                       // downshift so only 10 bits total are used
+	if words[13] != myskyDictionary[word1] {
+		return seed, errors.New("checksum does not match")
+	}
+	word2 := uint32(checksum[1]) << 10 // Load high so we can clear already used bits
+	word2 &= 0xffff                    // clear the 2 already used bits
+	word2 += uint32(checksum[2]) << 2  // load second word, also loaded high
+	word2 >>= 6                        // downshift so only 10 bits total are used
+	if words[14] != myskyDictionary[word2] {
+		return seed, errors.New("checksum does not match")
+	}
+
+	return seed, nil
+}
+
+// SeedToPhrase will return the phrase for a given mysky seed.
+func SeedToPhrase(seed Seed) string {
+	// Convert the seed to the first 13 words. This is copied from the
+	// skynet-js implementation.
+	curWord := 0
+	curBit := 0
+	wordBits := 10
+	var phrase string
+	for i := 0; i < 16; i++ {
+		nextByte := seed[i]
+		for j := 0; j < 8; j++ {
+			bitSet := (nextByte & (1 << (8 - j - 1))) > 0
+			if bitSet {
+				curWord |= 1 << (wordBits - curBit - 1)
+			}
+			curBit++
+			if curBit >= wordBits {
+				phrase += myskyDictionary[curWord]
+				phrase += " "
+				curWord = 0
+				curBit = 0
+				if i >= 14 {
+					wordBits = 8
+				}
+			}
+		}
+	}
+
+	// Determine the checksum of the phrase. This is copied from the
+	// skynet-js implementation.
+	checksum := sha512.Sum512(seed[:])
```

Comment:
This is not the same as `sha3.Sum512` right? But I guess this is what we want right?

I think it would be nice to have a test with a hardcoded seed which was created in mysky to make sure the go code is producing the same results.